File: algorithms/depth_first_search.java
// Depth First Search (DFS) is a graph traversal algorithm that explores as far as possible along each branch before backtracking.
// It is often used to traverse or search through trees or graphs.

// Consider a scenario where you are trying to navigate through a maze to find the exit.
// In this maze, you have multiple paths to choose from at each intersection.
// Depth First Search would involve exploring one path as far as possible before backtracking and exploring another path.

import java.util.*;

class Graph {
    private int V; // Number of vertices
    private LinkedList<Integer> adj[]; // Adjacency list representation

    Graph(int v) {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; i++) {
            adj[i] = new LinkedList();
        }
    }

    void addEdge(int v, int w) {
        adj[v].add(w);
    }

    // Depth First Search traversal starting from given vertex
    void DFSUtil(int v, boolean visited[]) {
        // Mark the current vertex as visited
        visited[v] = true;
        System.out.print(v + " ");

        // Recur for all vertices adjacent to this vertex
        Iterator<Integer> it = adj[v].iterator();
        while (it.hasNext()) {
            int n = it.next();
            if (!visited[n]) {
                DFSUtil(n, visited);
            }
        }
    }

    // Function to perform Depth First Search traversal
    void DFS(int v) {
        // Mark all the vertices as not visited(set as false by default in Java)
        boolean visited[] = new boolean[V];

        DFSUtil(v, visited);
    }
}

File: algorithms/depth-first_search.kt
// Depth-First Search (DFS) is an algorithm used to traverse or search tree or graph data structures.
// The algorithm starts at the root node (or an arbitrary node for graphs) and explores as far as possible along each branch before backtracking.

// Supporse you're exploring a cave network. You start at the entrance and at each junction, you choose a tunnel to explore.
// You go as deep as you can, backtracking only when you reach a dead end.
// In computer science, this approach of exhaustive exploration is askin to the Depth-First Search algorithm.

class Graph(private val numVertices: Int) {
    private val adjList: MutableList<MutableList<Int>> = ArrayList()

    init {
        for (i in 0 until numVertices) {
            adjList.add(mutableListOf())
        }
    }

    fun addEdge(v: Int, w: Int) {
        adjList[v].add(w)
    }

    private fun DFSUtil(v: Int, visited: BooleanArray) {
        visited[v] = true
        print("$v ")

        for (i in adjList[v]) {
            if (!visited[i]) {
                DFSUtil(i, visited)
            }
        }
    }

    fun DFS(startingVertex: Int) {
        val visited = BooleanArray(numVertices)
        DFSUtil(startingVertex, visited)
    }
}

fun main() {
    val graph = Graph(4)

    graph.addEdge(0, 1)
    graph.addEdge(0, 2)
    graph.addEdge(1, 2)
    graph.addEdge(2, 0)
    graph.addEdge(2, 3)
    graph.addEdge(3, 3)

    println("Depth First Traversal starting from vertex 2:")
    graph.DFS(2)
}
